hudi-common/src/main/java/org/apache/hudi/common/util/collection/ExternalSpillableMap.java

```diff
@@ -213,6 +213,8 @@ public R put(T key, R value) {
 
     if (this.inMemoryMap.containsKey(key)) {
       this.inMemoryMap.put(key, value);
+    } else if (inDiskContainsKey(key)) {
```

Comment:
Thanks for the fix @llincc . 

This is definitely one option of keeping subsequent versions of the same record (identified by key) in disk if the first version goes to disk. 

The other option would be : if the record was earlier in disk but based on current sizing constraints, we allow the new version of the record in in memory map but only after removing the key from diskBasedMap. 

There is difference in performance between the two option. In the second option, subsequent get() calls on those records will be faster. 

Thoughts ?

Reply:
@bvaradar 
thanks for your suggestions. this obviously has better performance when calling get() for option two. There are two ways to implement this: 

1.  Remove the old version record from disk first and then put the new record to in-memory map. but there is  a performance problem to remove record.

2. keep the behavior of putting data and deduplicate data when iterator.   But it is hard to implement hasNext() function of IteratorWrapper , do you have any good ideas about it ?

Reply:
@llincc : Implementation (1) should be good. For BitCaskDiskMap, remove() doesn't touch disk. Also, unlike get(), this scenario of evicting from disk and landing into memory is relatively bounded. So, lets go with implementation (1) option.